### PR TITLE
labhub: Add ban command

### PR DESCRIPTION
**Pull request by nvzard**
Add command to ban a user from all
rooms at once.

Closes https://github.com/coala/corobo/issues/549

### CODE REVIEW

File: plugins/labhub.py [hunk 1]
```diff
@@ -381,3 +382,43 @@ def pr_stats(self, msg, match):
...
+        joined_rooms = self.bot_config.ROOMS_TO_JOIN
+        headers = {'Content-Type': 'application/json',
+                   'Accept': 'application/json',
+                   'Authorization': 'Bearer ' +
+                   self.bot_config.BOT_IDENTITY['token']
+                   }
+        data = json.dumps({"username": sinner})
+
+        r = requests.get('https://api.gitter.im/v1/rooms', headers=headers)
+        room_data = json.loads(r.text)
```
Comment (meetmangukiya) [inline plugins/labhub.py]: You can use `r.json()` as well

File: plugins/labhub.py [hunk 2]
```diff
@@ -381,3 +382,43 @@ def pr_stats(self, msg, match):
...
+        if not self.TEAMS[self.GH_ORG_NAME + ' maintainers'].is_member(user):
+            yield 'Only maintainers are allowed to use this command.'
+            return
+
+        sinner = args[0]
+        if sinner.startswith('@'):
+            sinner = sinner[1:]
+
+        joined_rooms = self.bot_config.ROOMS_TO_JOIN
+        headers = {'Content-Type': 'application/json',
```
Comment (meetmangukiya) [inline plugins/labhub.py]: use this style:
```
dict = {
    key: value,
    key: value,
}
```

File: plugins/labhub.py [hunk 3]
```diff
@@ -381,3 +382,43 @@ def pr_stats(self, msg, match):
...
+        for room in joined_rooms:
+            for rm in room_data:
+                try:
+                    if rm['uri'] == room:
+                        url = 'https://api.gitter.im/v1/rooms/' + \
+                            rm['id'] + '/bans'
+                        r = requests.post(url, data=data, headers=headers)
+                        banned_rooms.append(room)
+                except KeyError:
+                    pass
```
Comment (meetmangukiya) [inline plugins/labhub.py]:
```
for room in filter(lambda x: x['uri'] in joined_rooms, room_data):
   rq = requests.post('https://api.gitter.im/v1/rooms/' + room['id'] + '/bans', data=data, headers=headers)
   banned_users.append(room['uri'])
```
Use something like this, shorter and better ^^ . Not including the exception handling bit, add it ;)

Comment (nvzard) [reply plugins/labhub.py]: sweet

**Comment by nvzard** [comment]
@meetmangukiya I didn't used vcr.py to test this because I didn't understand how to deal with the gitter token :/

File: config.py [hunk 4]
```diff
@@ -104,6 +104,7 @@
 
 ACCESS_CONTROLS = {'render test': {
     'allowrooms': ('coala/cobot-test', 'coala/corobo',)},
-    'LabHub:*': {'allowprivate': False}}
+    'LabHub:*': {'allowprivate': False}
```
Comment (jayvdb) [inline config.py]: so this is a syntax error.
not tested.
and something is wrong with CI that it didnt find this problem for you.
please fix.  separate PR.

File: plugins/ban.plug [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[Core]
+name = Ban
+module = ban
+
+[Documentation]
+description = Ban from all rooms at once
+[Python]
+version = 3
+
+[Errbot]
```
Comment (jayvdb) [inline plugins/ban.plug]: do we need this empty section?

if so, can we get a comment on it explaining why?

File: tests/ban_test.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    @patch('plugins.ban.json')
+    def test_ban_cmd(self, mockjson, mockreq):
+        ban, testbot = plugin_testbot(plugins.ban.Ban, logging.ERROR)
+        ban.activate()
+
+        ban.bot_config.ROOMS_TO_JOIN = ('coala/coala', 'coala/coala-bears')
+        ban.bot_config.BOT_IDENTITY['token'] = 'mocked?'
+
+        fake_room_data = [{'id': '130', 'uri': 'coala/coala'},
+        		  {'id': '234', 'name': 'Nitanshu'},
```
Comment (jayvdb) [inline tests/ban_test.py]: horrific style.
please get pyflakes active on these files.

Comment (nvzard) [reply tests/ban_test.py]: Sorry, that was looking regularly indented in vim.

File: plugins/ban.plug (hunk 5 above)
Comment (nvzard) [reply plugins/ban.plug]: Some of the other plugins' `plug` files were using it, I just borrowed their plug file as a template.
https://github.com/coala/corobo/blob/master/plugins/answer.plug
https://github.com/coala/corobo/blob/master/plugins/explain.plug

File: plugins/ban.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        corobo ban <@username/username>
+        """
+        user = msg.frm.nick
+        sinner = args[0]
+
+        if sinner.startswith('@'):
+            sinner = sinner[1:]
+
+        joined_rooms = self.bot_config.ROOMS_TO_JOIN
+        headers = {
```
Comment (meetmangukiya) [inline plugins/ban.py]: Use four level indentation, and the closing brace should be in same column as `h` in `headers`

File: tests/ban_test.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    @patch('plugins.ban.requests')
+    @patch('plugins.ban.json')
+    def test_ban_cmd(self, mockjson, mockreq):
+        ban, testbot = plugin_testbot(plugins.ban.Ban, logging.ERROR)
+        ban.activate()
+
+        ban.bot_config.ROOMS_TO_JOIN = ('coala/coala', 'coala/coala-bears')
+        ban.bot_config.BOT_IDENTITY['token'] = 'mocked?'
+
+        fake_room_data = [{'id': '130', 'uri': 'coala/coala'},
```
Comment (meetmangukiya) [inline tests/ban_test.py]: Use this style please 
```
a = [
    {...},
    {...},
...
]
```

File: plugins/ban.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            'Content-Type': 'application/json',
+            'Accept': 'application/json',
+            'Authorization': 'Bearer ' +
+            self.bot_config.BOT_IDENTITY['token']
```
Comment (meetmangukiya) [inline plugins/ban.py]: I think this can go above, and still be under 80 cols, check it

File: tests/ban_test.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    @patch('plugins.ban.json')
+    def test_ban_cmd(self, mockjson, mockreq):
+        ban, testbot = plugin_testbot(plugins.ban.Ban, logging.ERROR)
+        ban.activate()
+
+        ban.bot_config.ROOMS_TO_JOIN = ('coala/coala', 'coala/coala-bears')
+        ban.bot_config.BOT_IDENTITY['token'] = 'mocked?'
+
+        fake_room_data = [
+                {'id': '130', 'uri': 'coala/coala'},
```
Comment (meetmangukiya) [inline tests/ban_test.py]: why 8 spaces :confused: ?

File: plugins/ban.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+            sinner = sinner[1:]
+
+        joined_rooms = self.bot_config.ROOMS_TO_JOIN
+        headers = {
+            'Content-Type': 'application/json',
+            'Accept': 'application/json',
+            'Authorization': 'Bearer ' + self.bot_config.BOT_IDENTITY['token']
+        }
+
+        r = requests.get('https://api.gitter.im/v1/rooms', headers=headers)
```
Comment (meetmangukiya) [inline plugins/ban.py]: `delete` ?

Comment (nvzard) [reply plugins/ban.py]: Not it is necessary to get the room id. I have used `requests.delete` in on line #75

Comment (meetmangukiya) [reply plugins/ban.py]: oh right, misread

**Comment by meetmangukiya** [comment]
ack https://github.com/coala/corobo/pull/553/commits/36e8b5ad6c2e7b2f73f2b40f911c41551c8253a0

**Comment by meetmangukiya** [comment]
@gitmate-bot rebase

**Comment by meetmangukiya** [comment]
ack https://github.com/coala/corobo/pull/553/commits/e24d461af968ef2bce95130c3bab12364dd962e1

**Comment by meetmangukiya** [comment]
@gitmate-bot ff

**Pull request merged**